extensions/amp-powr-player/0.1/amp-powr-player.js

```diff
@@ -0,0 +1,432 @@
+    this.readyTimeout_ = null;
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+
+    /**@private {?string} */
+    this.playerId_ = null;
+
+    /** @private {?../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacements_ = null;
+  }
+
+  /** @override */
+  preconnectCallback() {
+    this.preconnect.url('https://player.powr.com');
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  viewportCallback(visible) {
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {
+      visible,
+    });
+  }
+
+  /** @override */
+  buildCallback() {
+    const ampdoc = this.getAmpDoc();
+    const deferred = new Deferred();
+
+    this.urlReplacements_ = Services.urlReplacementsForDoc(ampdoc);
+    this.playerReadyPromise_ = deferred.promise;
+    this.playerReadyResolver_ = deferred.resolve;
+
+    // Warn if the player does not have video interface support
+    this.readyTimeout_ = Services.timerFor(window).delay(() => {
+      user().warn(TAG,
+          'Did not receive ready callback from player %s.' +
+        ' Ensure it has the iframe plugin.', this.playerId_);
+    }, 3000);
+
+    this.playerReadyResolver_(this.iframe_);
+  }
+
+  /** @override */
+  layoutCallback() {
+    const iframe = createFrameFor(this, this.getIframeSrc_());
+
+    this.iframe_ = iframe;
+
+    this.unlistenMessage_ = listen(
+        this.win,
+        'message',
+        e => this.handlePlayerMessage_(e));
+
+    return this.loadPromise(iframe)
+        .then(() => this.playerReadyPromise_);
+  }
+
+  /**
+   * Sends a command to the player through postMessage.
+   * @param {string} command
+   * @param {*=} arg
+   * @private
+   * */
+  sendCommand_(command, arg) {
+    this.playerReadyPromise_.then(() => {
+      // We still need to check this.iframe_ as the component may have
+      // been unlaid out by now.
+      if (this.iframe_ && this.iframe_.contentWindow) {
+        this.iframe_.contentWindow. /*OK*/ postMessage(JSON.stringify(dict({
+          'command': command,
+          'args': arg,
+        })), 'https://player.powr.com');
+      }
+    });
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  handlePlayerMessage_(event) {
+    const {element} = this;
+
+    if (event.source != this.iframe_.contentWindow) {
+      return;
+    }
+
+    const eventData = getData(event);
+    if (!isJsonOrObj(eventData)) {
+      return;
+    }
+
+    const data = objOrParseJson(eventData);
+    if (data === undefined) {
+      return; // We only process valid JSON.
+    }
+
+    const eventType = data['event'];
+    if (!eventType) {
+      return;
+    }
+
+    if (eventType === 'ready') {
+      this.onReady_(data);
+    }
+
+    if (eventType === 'playing') {
+      this.playing_ = true;
+    }
+    if (eventType === 'pause') {
+      this.playing_ = false;
+    }
+
+    if (redispatch(element, eventType, {
```

Comment:
nit: define as top-level constant so the dictionary doesn't get recreated each time a message gets received.